edb/server/protocol/auth_ext/base.py

```diff
@@ -48,3 +53,73 @@ async def fetch_user_info(self, token: str) -> data.UserInfo:
 
     def _maybe_isoformat_to_timestamp(self, value: str | None) -> float | None:
         return datetime.fromisoformat(value).timestamp() if value else None
+
+
+class OpenIDProvider(BaseProvider):
+    def __init__(self, name: str, issuer_url: str, *args, **kwargs):
+        super().__init__(name, issuer_url, *args, **kwargs)
+
+    async def get_code_url(self, state: str, redirect_uri: str) -> str:
+        oidc_config = await self._get_oidc_config()
+        params = {
+            "client_id": self.client_id,
+            "scope": "openid profile email",
+            "state": state,
+            "redirect_uri": redirect_uri,
+            "nonce": str(uuid.uuid4()),
+            "response_type": "code",
+        }
+        encoded = urllib.parse.urlencode(params)
+        return f"{oidc_config.authorization_endpoint}?{encoded}"
+
+    async def exchange_code(self, code: str) -> str:
+        oidc_config = await self._get_oidc_config()
+        data = {
+            "grant_type": "authorization_code",
+            "code": code,
+            "client_id": self.client_id,
+            "client_secret": self.client_secret,
+        }
+
+        token_endpoint = urllib.parse.urlparse(oidc_config.token_endpoint)
+        async with self.http_factory(
+            base_url=f"{token_endpoint.scheme}://{token_endpoint.netloc}"
+        ) as client:
+            resp = await client.post(
+                token_endpoint.path,
+                json=data,
+            )
+            token = resp.json()["id_token"]
```

Comment:
I'm a little meh about this, but basically:

OpenID connect returns an `id_token` which contains the data we care about for the "user info" already. So instead of returning the authorization token and making a request to the user info endpoint, I am just returning the `id_token` and extracting the user info from the claims.

I considered refactoring this to return a tuple of both tokens, does that seem cleaner? I feel like we might want that once we want to start returning the authorization token to the user, but I didn't want to prematurely complicate the code if there was another/better way to approach this.

Reply:
Alright, I decided that I can't live with this how it is, so refactored it a bunch in [`d995fef` (#6013)](https://github.com/edgedb/edgedb/pull/6013/commits/d995fef386e7e36027f67a7fce899e882b8ec714)